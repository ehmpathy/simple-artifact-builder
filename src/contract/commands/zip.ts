import { Command, flags } from '@oclif/command';
import { apply } from '../../logic/commands/apply';

export default class Apply extends Command {
  public static description =
    "apply fixes to all files which have failed to adhere to any of the project's declared practices and have an automatic fix available.";

  public static flags = {
    help: flags.help({ char: 'h' }),
    config: flags.string({
      char: 'c',
      description: 'path to the artifact config yml',
      required: true,
      default: 'artifact.yml',
    }),
  };

  public async run() {
    const { flags } = this.parse(Apply);
    const config = flags.config!;

    // define config path
    const configPath = config.slice(0, 1) === '/' ? config : `${process.cwd()}/${config}`; // if starts with /, consider it as an absolute path
    await apply({ configPath, });
  }
}
