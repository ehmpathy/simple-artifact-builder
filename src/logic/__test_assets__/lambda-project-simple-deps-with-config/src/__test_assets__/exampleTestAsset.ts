export const asset = () => console.log('hello');
