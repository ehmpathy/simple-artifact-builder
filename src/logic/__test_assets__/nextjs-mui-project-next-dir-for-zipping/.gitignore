.artifact/contents.zip
.artifact/contents.sizes.json
