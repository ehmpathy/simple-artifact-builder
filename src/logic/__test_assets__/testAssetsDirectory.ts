export const TEST_ASSETS_DIRECTORY = __dirname;
