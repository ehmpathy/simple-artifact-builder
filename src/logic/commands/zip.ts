import chalk from 'chalk';

import { readConfig } from '../config/readConfig';
import { zipArtifactContents } from '../containers/zipArtifactContents';
import { buildArtifactContents } from '../contents/buildArtifactContents';

export const zip = async ({
  absoluteConfigPath,
}: {
  absoluteConfigPath: string;
}) => {
  // read the config
  console.log('🔎 reading config...'); // tslint:disable-line: no-console
  const { config, projectRootDirectory } = await readConfig({
    absoluteConfigPath,
  });

  // build the artifact
  console.log('🔨 building artifact...'); // tslint:disable-line: no-console
  await buildArtifactContents({ projectRootDirectory, config });

  // zip the artifact
  console.log('🗜️  zipping artifact...'); // tslint:disable-line: no-console
  await zipArtifactContents({ projectRootDirectory });

  console.log(`🎉 done! ${chalk.grey('(.artifact/contents.zip)')}`); // tslint:disable-line: no-console
};
