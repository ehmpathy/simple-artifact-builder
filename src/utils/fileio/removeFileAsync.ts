import fs from 'fs';
import util from 'util';

const unlink = util.promisify(fs.unlink);

export const removeFileAsync = async ({ path }: { path: string }) =>
  unlink(path);
