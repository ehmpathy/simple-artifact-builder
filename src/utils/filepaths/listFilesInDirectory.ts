import glob from 'fast-glob';

/**
 * returns paths to all files inside this directory or its nested directories
 */
export const listFilesInDirectory = async ({
  directory,
}: {
  directory: string;
}) => glob('**/*', { cwd: directory, dot: true, onlyFiles: true });
