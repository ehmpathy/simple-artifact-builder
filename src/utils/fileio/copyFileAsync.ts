import fs from 'fs';
import util from 'util';

import { getDirOfPath } from '../filepaths/getDirOfPath';
import { makeDirectoryAsync } from './makeDirAsync';

const copyFile = util.promisify(fs.copyFile);

export const copyFileAsync = async ({
  from,
  to,
}: {
  from: string;
  to: string;
}) => {
  await makeDirectoryAsync({ directoryPath: await getDirOfPath(to) }); // find or create the dir, before writing the file to the dir
  await copyFile(from, to);
};
