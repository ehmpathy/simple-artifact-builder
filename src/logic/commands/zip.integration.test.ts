import { removeFileAsync } from '../../utils/fileio/removeFileAsync';
import { listFilesInDirectory } from '../../utils/filepaths/listFilesInDirectory';
import { log } from '../../utils/logger';
import { TEST_ASSETS_DIRECTORY } from '../__test_assets__/testAssetsDirectory';
import { zip } from './zip';

const logSpy = jest.spyOn(console, 'log').mockImplementation(() => log.debug); // swap to log debug so its not displaying during tests by default

describe('apply', () => {
  beforeEach(() => jest.clearAllMocks());
  it('should be able to apply for an example project', async () => {
    const projectRootDirectory = `${TEST_ASSETS_DIRECTORY}/lambda-project-simple-deps-with-config`;

    // delete the generated files to get the testing environment to a clean state
    const autoGeneratedFilesInDirectory = [
      '.artifact/contents.manifest.json',
      '.artifact/contents.sizes.json',
      '.artifact/contents.zip',
    ];
    const filesInDirectory = await listFilesInDirectory({ directory: projectRootDirectory });
    await Promise.all(
      filesInDirectory
        .filter((path) => autoGeneratedFilesInDirectory.includes(path))
        .map((path) => removeFileAsync({ path: `${projectRootDirectory}/${path}` })),
    );

    // now apply
    await zip({
      absoluteConfigPath: `${projectRootDirectory}/artifact.yml`,
    });
    expect(logSpy.mock.calls).toMatchSnapshot(); // log an example of what the user will see

    // and check that the autogenerated files now exist
    const filesInDirectoryNow = await listFilesInDirectory({ directory: projectRootDirectory });
    autoGeneratedFilesInDirectory.forEach((file) => expect(filesInDirectoryNow).toContain(file));
  });
});
