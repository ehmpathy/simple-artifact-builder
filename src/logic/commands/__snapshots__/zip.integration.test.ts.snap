// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply should be able to apply for an example project 1`] = `
[
  [
    "🔎 reading config...",
  ],
  [
    "🔨 building artifact...",
  ],
  [
    "🗜️  zipping artifact...",
  ],
  [
    "🎉 done! [90m(.artifact/contents.zip)[39m",
  ],
]
`;
